src/pages/Admin/Room/Edit/RoomEditModal.tsx

```diff
@@ -183,6 +183,24 @@ export const RoomEditModal: React.FC<RoomEditModalProps> = ({
             </div>
           </Form.Row>
 
+          <Form.Row>
+            <div className="room-edit-modal__input">
+              <Form.Label>Change room type</Form.Label>
+              <Form.Control
+                as="select"
+                custom
+                name="type"
+                ref={register}
+                defaultValue={room.type}
+              >
+                <option />
```

Comment:
I wonder if we should add a label like 'Default' or similar here, rather than just being an empty option? Might be better UX?

```suggestion
                <option value="">Default</option>
```

Reply:
Also, I'm not sure how this will translate into the data we send to the backend, and ultimately how it gets stored. Currently though, the 'default case' in the backend should be that the room type doesn't exist (which will translate to `undefined` in the app code). We don't want to set it to an empty string, etc. So we might need to tweak some things to ensure that works as expected.

Reply:
> I wonder if we should add a label like 'Default' or similar here, rather than just being an empty option? Might be better UX?

Using `<option>Default</option>` so that value is set to `undefined`.

Reply:
We should agree on a better way to handle this. If we follow this pattern, editing fields are going to be a pain. Would setting `null` work? Trying to figure out a convention which is easy to maintain.

Reply:
Seemingly based on a comment you made elsewhere, removing the value=“” is all that’s required?

Reply:
That is what I had expected but wasn't the case when I tested after posting the comment. I deleted the comment thereafter.

Reply:
Instead of just deleting the comment, you should correct what you said so that you’re communicating that new insight with others. That would have saved this back and forth; and me operating on incorrect information.

Reply:
> We should agree on a better way to handle this. If we follow this pattern, editing fields are going to be a pain. Would setting null work? Trying to figure out a convention which is easy to maintain.

Why are editing fields going to be a pain/etc if we do this? Would seem it would just be a couple of tiny changes to transform the data?

Reply:
The comment was deleted within seconds of posting it. My workflow does not involve reading review comments off Github. I realise that isn't the case with everyone here, will post updates going forward.

Reply:
> > We should agree on a better way to handle this. If we follow this pattern, editing fields are going to be a pain. Would setting null work? Trying to figure out a convention which is easy to maintain.
> 
> Why are editing fields going to be a pain/etc if we do this? Would seem it would just be a couple of tiny changes to transform the data?

Yes, the changes are trivial but they might introduce bugs later. The reasoning is; this field behaves differently than other fields or how developers perceive an app to work.
Editing a field, updates a value but changing that pattern here, we delete the key as well. This is an exception which may be overlooked by future devs. We may add comments and tests to have a guard against such bugs but it will be one more thing developers have to look throughout codebase for such exceptions.
Even on the access side, we will now have to check whether the key is defined before checking for match.

In SQL world, this would be equivalent to running a drop column every time we delete the value of a column. But we are in NoSQL land so maybe this is how it works.

IRL, will a room loose it's `type` if it does not have a `type` or does the room `type` become default? Answer to that question may help think about a solution too.

Happy to roll with a convention we pick which we can carry forward.

Reply:
The current default is that the ‘type’ key doesn’t exist, thus it is ‘undefined’ within the code when ‘read back’. The pattern that we tend to follow is that defaults aren’t defined in the database, they’re defined in the code. Handling this is a standard pattern, and quite trivial with TypeScripts ‘safe access’ checks (foo?.bar) and defaults (foo?.bar ?? “baz”)

Reply:
> Yes, the changes are trivial but they might introduce bugs later. The reasoning is; this field behaves differently than other fields or how developers perceive an app to work.
Editing a field, updates a value but changing that pattern here, we delete the key as well. This is an exception which may be overlooked by future devs.

Further to my comments in https://github.com/sparkletown/sparkle/pull/1305#discussion_r611932752, we 'delete the key' only because firebase doesn't let us explicitly set `undefined` as a key that gets stored on the backend. This is a pattern we already use in other locations within the app.

Reply:
If this is a pattern which is being followed, happy to roll with it. Maybe we should document these nuances somewhere more explicitly so other developers joining in are aware of the conventions.

My personal preference is to use `null` and try and avoid safe checks as they add additional conditional checks leading slower benchmarks. More like pre-optimization; and we all know what that leads to! 😃

Reply:
There is a PR on the public repo titled documentation that captures a lot of these nuances. There are also backlogged issues about documentation, as well as the internal documentation repo itself. These are all a WIP, but if you find an area lacking, you’re more than welcome to create a new backlog issue for it and/or write up some of the existing ‘snapshots’ of how we do things more formally. We eventually want to get to this, but time is limited, and there have been higher priorities.

Reply:
The massive increase in engineer efficiency trumps the negligible small gains we would get by avoiding using safe accessors/etc.

Let’s not hyperfocus on ‘code gold’ optimisation that is unlikely to have any real world impact on users. Particularly when that reduces engineer efficiency and/or code maintainability.

Reply:
@0xdevalias Issues aren't the best place to document these as new engineer aren't sure what to search for in the issues. Wiki/Google Docs might be a better option.

Totally agree on the `code golf` optimisation doesn't make sense. My primary concern was around the pattern which has been addresses as thats the convention here.

Reply:
As I said above, they aren’t permanently documented in issues, the issues are there to capture the insights to be written up into proper documentation within our ‘internal docs’ repo, as per our standards.

Reply:
@0xdevalias I couldn't lookup an example where we remove keys in `/api`. Help me locate an example which I can follow.

Reply:
The docs links above should show you how to do it. I’m not sure if removing the key is still in an ‘in review’ PR, probably is. Maybe look at the auditorium v2 one? Might be there.

You can probably also just search for Field Value or similar to find similar examples.

But basically, as I believe the linked docs describe, within an update, the ‘value’ is set to the special FieldValue entry that says to delete it.

This might be something that gets added on the backend function even, depending where you think it makes more sense to have it, though I believe that might require changes in our frontend validation if we do.

Reply:
@0xdevalias Please share a piece of code which shows the convention. I am not familiar enough with the codebase to dig it out.

Reply:
Does searching the codebase for the literal “FieldValue” (without the quotes; as mentioned above) not find it for you?

Reply:
No, I didn't. Most of them were array operations, none of them were about deleting a key from an object before adding back to an array.

Reply:
@mike-lvov Are you able to help out here by chance? Not sure I’m going to be able to explain better than I already have + the docs/references I’ve already included. Perhaps you’re able to do so better..

Reply:
@TroySK 

`export const setGridData = async ({
  venueId,
  userId,
  gridData,
}: SetGridDataProps): Promise<void> => {
  const userProfileRef = firebase.firestore().collection("users").doc(userId);

  const newGridData = {
    [`data.${venueId}`]: gridData ?? firebase.firestore.FieldValue.delete(),
  };

  return userProfileRef.update(newGridData).catch((err) => {
    Bugsnag.notify(err, (event) => {
      event.addMetadata("context", {
        location: "api/profile::setGridData",
        venueId,
        userId,
        gridData,
      });
    });
  });
};
`

Reply:
Thanks @mike-lvov! This is close but not exactly what I was looking for. Please review https://github.com/sparkletown/sparkle/pull/1305/commits/2c3bf9a8656a0c591ab4774e50ad1a9e8fb16e31. This is what I wanted to avoid as this kind of special cases lead to bugs in the future.

Reply:
@TroySK I don't understand the purpose/intent of the code in the commit you linked, and how that relates to the above? What @mike-lvov shared was what I was talking about in my original suggestions here.

Reply:
The pattern @mike-lvov attached doesn't work for what I am trying to solve. It could be my knowledge of TypeScript is lacking. `room`s are array and inserted using an upsert. The `type` key on the room object needs to be removed before adding it back to the array if it is null. Not sure how null coalescing can be used here. Are you suggesting I use `??` and use `delete room.type`? I selected this alternative as it doesn't mutate the object.

Reply:
> Are you suggesting I use `??` and use `delete room.type`? I selected this alternative as it doesn't mutate the object.

Assuming I am understanding what you're saying here correctly, that was what I was suggesting, yeah.

From @mike-lvov's screenshot above, this bit:

![image](https://user-images.githubusercontent.com/753891/115795876-18f5df80-a414-11eb-8f15-ab4b6a027d9d.png)

Use the data if we have data, or if it's `undefined` then we tell firebase to delete the field.